src/rust/engine/graph/src/lib.rs

```diff
@@ -988,25 +474,28 @@ impl<N: Node> Graph<N> {
   where
     C: NodeContext<Node = N>,
   {
-    // Get or create the destination, and then insert the dep and return its state.
-    let mut inner = self.inner.lock().unwrap();
-    let dst_id = {
-      // TODO: doing cycle detection under the lock... unfortunate, but probably unavoidable
-      // without a much more complicated algorithm.
-      let potential_dst_id = inner.ensure_entry(EntryKey::Valid(dst_node.clone()));
-      if inner.detect_cycle(src_id, potential_dst_id) {
-        // Cyclic dependency: declare a dependency on a copy of the Node that is marked Cyclic.
-        inner.ensure_entry(EntryKey::Cyclic(dst_node))
-      } else {
-        // Valid dependency.
-        potential_dst_id
-      }
+    let (maybe_entry, entry_id) = {
+      // Get or create the destination, and then insert the dep and return its state.
+      let mut inner = self.inner.lock().unwrap();
+      let dst_id = {
+        // TODO: doing cycle detection under the lock... unfortunate, but probably unavoidable
+        // without a much more complicated algorithm.
+        let potential_dst_id = inner.ensure_entry(EntryKey::Valid(dst_node.clone()));
+        if inner.detect_cycle(src_id, potential_dst_id) {
+          // Cyclic dependency: declare a dependency on a copy of the Node that is marked Cyclic.
+          inner.ensure_entry(EntryKey::Cyclic(dst_node))
+        } else {
+          // Valid dependency.
+          potential_dst_id
+        }
+      };
+      inner.pg.add_edge(src_id, dst_id, ());
+      (inner.entry_for_id(dst_id).cloned(), dst_id)
     };
 
     // Declare the dep, and return the state of the destination.
-    inner.pg.add_edge(src_id, dst_id, ());
-    if let Some(entry) = inner.entry_for_id_mut(dst_id) {
-      entry.get(context, dst_id).map(|(res, _)| res).to_boxed()
+    if let Some(mut entry) = maybe_entry {
+      entry.get(context, entry_id).map(|(res, _)| res).to_boxed()
```

Comment:
I'm kind of tempted to make it so that `context` is pre-cloned with the correct `entry_id` here because it feels weird needing to pass both around, but it looks slightly more complicated than I would have guessed, and also probably means we end up cloning `Context`s more than we strictly need to... WDYT?

Reply:
I have wobbled back and forth on this a bit, yea. IIRC, the only case where the context is not cloned inside `Entry::get` is when the dependency is cyclic? And that should be quite rare.